File: quantura_android/app/src/main/java/com/quantura/quanturaapp/ads/AppOpenAdManager.kt
package com.quantura.quanturaapp.ads

import android.app.Activity
import android.app.Application
import android.os.Handler
import android.os.Looper
import android.os.SystemClock
import android.util.Log
import androidx.lifecycle.DefaultLifecycleObserver
import androidx.lifecycle.LifecycleOwner
import androidx.lifecycle.ProcessLifecycleOwner
import com.google.android.gms.ads.AdRequest
import com.google.android.gms.ads.FullScreenContentCallback
import com.google.android.gms.ads.LoadAdError
import com.google.android.gms.ads.appopen.AppOpenAd
import com.quantura.quanturaapp.MainActivity
import com.quantura.quanturaapp.config.AdFormat
import com.quantura.quanturaapp.config.AdPlatform
import com.quantura.quanturaapp.config.RemoteConfigManager

class AppOpenAdManager(
    private val application: Application,
    private val remoteConfigManager: RemoteConfigManager,
    private val adManager: AdManager,
) : Application.ActivityLifecycleCallbacks, DefaultLifecycleObserver {
    private val tag = "AppOpenAdManager"
    private val maxCacheAgeMs = 4 * 60 * 60 * 1000L
    private val mainHandler = Handler(Looper.getMainLooper())

    @Volatile
    private var appOpenAd: AppOpenAd? = null
    @Volatile
    private var loadedAdUnitId: String = ""
    @Volatile
    private var isLoadingAd = false
    @Volatile
    private var isShowingAd = false
    @Volatile
    private var loadedAtMs = 0L
    @Volatile
    private var currentActivity: Activity? = null
    @Volatile
    private var currentActivityResumed = false
    @Volatile
    private var presentationBlockedByAuthGate = false
    @Volatile
    private var canPresentInCurrentForeground = true

    fun start() {
        application.registerActivityLifecycleCallbacks(this)
        ProcessLifecycleOwner.get().lifecycle.addObserver(this)
    }

    override fun onStart(owner: LifecycleOwner) {
        canPresentInCurrentForeground = true
        showAdIfAvailable()
    }

    override fun onStop(owner: LifecycleOwner) {
        canPresentInCurrentForeground = true
    }

    fun setPresentationBlockedByAuthGate(blocked: Boolean) {
        presentationBlockedByAuthGate = blocked
        if (blocked) {
            Log.d(tag, "Auth gate is visible; app open presentation is deferred.")
        }
    }

    fun loadAdIfNeeded() {
        if (!remoteConfigManager.areAdsEnabled()) {
            Log.d(tag, "App open ads disabled by feature flag.")
            AdDebugStatusRegistry.updateLoad("app_open", "disabled")
            return
        }
        if (!remoteConfigManager.isAdFormatEnabled(platform = AdPlatform.ANDROID, format = AdFormat.APP_OPEN)) {
            Log.d(tag, "App open ads disabled by format flag.")
            AdDebugStatusRegistry.updateLoad("app_open", "disabled:format_off")
            appOpenAd = null
            loadedAdUnitId = ""
            loadedAtMs = 0L
            return
        }
        if (!MobileAdsBootstrap.isInitialized()) {
            Log.d(tag, "App open load deferred until Mobile Ads finishes initializing.")
            AdDebugStatusRegistry.updateLoad("app_open", "waiting:sdk_init")
            MobileAdsBootstrap.runWhenInitialized {
                mainHandler.post { loadAdIfNeeded() }
            }
            return
        }
        if (isLoadingAd || isAdAvailable()) return
        val unitId = remoteConfigManager.resolveAdUnitId(
            platform = AdPlatform.ANDROID,
            format = AdFormat.APP_OPEN
        )
        Log.d(tag, "Loading app open ad unit=$unitId")
        AdDebugStatusRegistry.updateLoad("app_open", "loading")
        isLoadingAd = true
        AppOpenAd.load(
            application,
            unitId,
            AdRequest.Builder().build(),
            object : AppOpenAd.AppOpenAdLoadCallback() {
                override fun onAdLoaded(ad: AppOpenAd) {
                    isLoadingAd = false
                    val desiredUnitId = remoteConfigManager.resolveAdUnitId(
                        platform = AdPlatform.ANDROID,
                        format = AdFormat.APP_OPEN
                    )
                    if (unitId != desiredUnitId) {
                        appOpenAd = null
                        loadedAdUnitId = ""
                        loadedAtMs = 0L
                        Log.i(
                            tag,
                            "[Ads][Android] Discarding stale app open load oldUnit=$unitId newUnit=$desiredUnitId"
                        )
                        loadAdIfNeeded()
                        return
                    }
                    appOpenAd = ad
                    loadedAdUnitId = unitId
                    loadedAtMs = SystemClock.elapsedRealtime()
                    Log.d(tag, "App open ad load succeeded.")
                    Log.i(tag, "[Ads][Android] Load success for app_open")
                    AdDebugStatusRegistry.updateLoad("app_open", "loaded")
                    // If app is already foregrounded, attempt an immediate presentation.
                    showAdIfAvailable()
                }

                override fun onAdFailedToLoad(loadAdError: LoadAdError) {
                    isLoadingAd = false
                    appOpenAd = null
                    loadedAdUnitId = ""
                    loadedAtMs = 0L
                    Log.w(tag, "App open ad load failed: ${loadAdError.message}")
                    Log.w(tag, "[Ads][Android] Load fail for app_open: ${loadAdError.message}")
                    AdDebugStatusRegistry.updateLoad("app_open", "failed:${loadAdError.message}")
                }
            }
        )
    }

    fun showAdIfAvailable() {
        if (!remoteConfigManager.areAdsEnabled()) return
        if (isShowingAd) return
        if (!remoteConfigManager.isAdFormatEnabled(platform = AdPlatform.ANDROID, format = AdFormat.APP_OPEN)) {
            Log.d(tag, "Skipping app open show; format flag is disabled.")
            AdDebugStatusRegistry.updateShow("app_open", "skipped:format_off")
            return
        }
        if (!canPresentInCurrentForeground) {
            Log.d(tag, "Skipping app open show; already presented in current foreground.")
            AdDebugStatusRegistry.updateShow("app_open", "skipped:already_presented")
            return
        }
        if (presentationBlockedByAuthGate) {
            Log.d(tag, "Skipping app open show; auth gate is visible.")
            AdDebugStatusRegistry.updateShow("app_open", "skipped:auth_gate")
            return
        }
        if (adManager.isShowingFullScreenAd()) {
            Log.d(tag, "Skipping app open show; another fullscreen ad is visible.")
            AdDebugStatusRegistry.updateShow("app_open", "skipped:fullscreen_visible")
            return
        }

        val activity = currentActivity
        if (!canShowOn(activity)) {
            Log.d(tag, "Skipping app open show; activity not ready (${activity?.javaClass?.simpleName ?: "none"}).")
            AdDebugStatusRegistry.updateShow("app_open", "skipped:activity_not_ready")
            return
        }
        val presenterActivity = activity ?: return

        val cached = appOpenAd
        if (!isAdAvailable() || cached == null) {
            Log.d(tag, "No fresh app open ad available; loading.")
            AdDebugStatusRegistry.updateShow("app_open", "skipped:not_ready")
            loadAdIfNeeded()
            return
        }

        isShowingAd = true
        canPresentInCurrentForeground = false
        cached.fullScreenContentCallback = object : FullScreenContentCallback() {
            override fun onAdShowedFullScreenContent() {
                Log.d(tag, "App open ad shown.")
                Log.i(tag, "[Ads][Android] Show success for app_open")
                AdDebugStatusRegistry.updateShow("app_open", "shown")
            }

            override fun onAdImpression() {
                val adUnitId = remoteConfigManager.resolveAdUnitId(
                    platform = AdPlatform.ANDROID,
                    format = AdFormat.APP_OPEN
                )
                AdImpressionReporter.report(
                    context = application.applicationContext,
                    adFormat = "app_open",
                    adUnitId = adUnitId,
                    placement = "app_open"
                )
            }

            override fun onAdFailedToShowFullScreenContent(adError: com.google.android.gms.ads.AdError) {
                Log.w(tag, "App open ad failed to show: ${adError.message}")
                isShowingAd = false
                canPresentInCurrentForeground = true
                appOpenAd = null
                loadedAtMs = 0L
                AdDebugStatusRegistry.updateShow("app_open", "failed:${adError.message}")
                loadAdIfNeeded()
            }

            override fun onAdDismissedFullScreenContent() {
                Log.d(tag, "App open ad dismissed.")
                isShowingAd = false
                appOpenAd = null
                loadedAtMs = 0L
                AdDebugStatusRegistry.updateShow("app_open", "dismissed")
                loadAdIfNeeded()
            }
        }
        Log.d(tag, "Presenting app open ad.")
        try {
            cached.show(presenterActivity)
        } catch (error: Exception) {
            isShowingAd = false
            canPresentInCurrentForeground = true
            appOpenAd = null
            loadedAtMs = 0L
            Log.w(tag, "App open ad show threw exception: ${error.message}")
            AdDebugStatusRegistry.updateShow("app_open", "failed:${error.message}")
            loadAdIfNeeded()
        }
    }

    fun refreshForRemoteConfigChange() {
        if (!remoteConfigManager.areAdsEnabled() || !remoteConfigManager.isAdFormatEnabled(
                platform = AdPlatform.ANDROID,
                format = AdFormat.APP_OPEN
            )
        ) {
            appOpenAd = null
            loadedAdUnitId = ""
            loadedAtMs = 0L
            return
        }
        val desiredUnitId = remoteConfigManager.resolveAdUnitId(
            platform = AdPlatform.ANDROID,
            format = AdFormat.APP_OPEN
        )
        if (appOpenAd != null && loadedAdUnitId != desiredUnitId) {
            Log.i(
                tag,
                "[Ads][Android] Clearing cached app open after RC change oldUnit=$loadedAdUnitId newUnit=$desiredUnitId"
            )
            appOpenAd = null
            loadedAdUnitId = ""
            loadedAtMs = 0L
        }
    }

    fun debugState(): AdFormatStatusSnapshot {
        return AdDebugStatusRegistry.snapshot(listOf("app_open")).first()
    }

    private fun isAdAvailable(): Boolean {
        if (appOpenAd == null) return false
        val age = SystemClock.elapsedRealtime() - loadedAtMs
        return age in 0..<maxCacheAgeMs
    }

    private fun canShowOn(activity: Activity?): Boolean {
        if (activity == null) return false
        if (activity !is MainActivity) return false
        if (activity.isFinishing || activity.isDestroyed) return false
        if (!currentActivityResumed) return false
        return true
    }

    override fun onActivityCreated(activity: Activity, savedInstanceState: android.os.Bundle?) {}

    override fun onActivityStarted(activity: Activity) {
        currentActivity = activity
    }

    override fun onActivityResumed(activity: Activity) {
        currentActivity = activity
        currentActivityResumed = true
        if (activity is MainActivity) {
            showAdIfAvailable()
        }
    }

    override fun onActivityPaused(activity: Activity) {
        if (currentActivity === activity) {
            currentActivityResumed = false
        }
    }

    override fun onActivityStopped(activity: Activity) {}

    override fun onActivitySaveInstanceState(activity: Activity, outState: android.os.Bundle) {}

    override fun onActivityDestroyed(activity: Activity) {
        if (currentActivity === activity) {
            currentActivity = null
            currentActivityResumed = false
            isShowingAd = false
        }
    }
}

File: quantura_ios/quantura_ios/AppOpenAdManager.swift
import Foundation
#if canImport(GoogleMobileAds) && canImport(UIKit)
import GoogleMobileAds
import UIKit

@MainActor
final class AppOpenAdManager: NSObject, FullScreenContentDelegate {
    private let remoteConfigManager: RemoteConfigManager
    private let adManager: AdManager
    private var appOpenAd: AppOpenAd?
    private var isLoading = false
    private var isShowing = false
    private var wantsForegroundPresentation = false
    private var waitingForSdkStart = false
    private var presentationBlockedByAuthGate = false
    private var pendingShowRetryTask: Task<Void, Never>?
    private var loadedAt: Date?
    private let maxCacheAgeSeconds: TimeInterval = 4 * 60 * 60

    init(remoteConfigManager: RemoteConfigManager, adManager: AdManager) {
        self.remoteConfigManager = remoteConfigManager
        self.adManager = adManager
    }

    func preloadAdIfNeeded() {
        guard remoteConfigManager.areAdsEnabled() else {
            print("[Ads][iOS][AppOpen] Ads disabled by feature flag.")
            AdDebugStatusStore.shared.updateLoad(format: "app_open", status: "disabled")
            return
        }
        guard remoteConfigManager.isAdFormatEnabled(format: .appOpen) else {
            print("[Ads][iOS][AppOpen] App open disabled by format flag.")
            AdDebugStatusStore.shared.updateLoad(format: "app_open", status: "disabled:format_off")
            appOpenAd = nil
            loadedAt = nil
            return
        }
        guard MobileAdsBootstrap.shared.isReady else {
            AdDebugStatusStore.shared.updateLoad(format: "app_open", status: "waiting:sdk_init")
            guard !waitingForSdkStart else { return }
            waitingForSdkStart = true
            MobileAdsBootstrap.shared.whenReady { [weak self] success in
                guard let self else { return }
                self.waitingForSdkStart = false
                guard success else { return }
                self.preloadAdIfNeeded()
            }
            return
        }
        guard !isLoading else { return }
        guard !isAdFresh else { return }
        loadAd()
    }

    func sceneDidBecomeActive() {
        print("[Ads][iOS][AppOpen] Scene became active; eligible for foreground presentation.")
        wantsForegroundPresentation = true
        pendingShowRetryTask?.cancel()
        pendingShowRetryTask = nil
        showIfAvailable()
    }

    func setPresentationBlockedByAuthGate(_ blocked: Bool) {
        let wasBlocked = presentationBlockedByAuthGate
        presentationBlockedByAuthGate = blocked
        if blocked {
            print("[Ads][iOS][AppOpen] Presentation blocked by auth gate.")
        } else if wasBlocked {
            showIfAvailable()
        }
    }

    private var isAdFresh: Bool {
        guard appOpenAd != nil, let loadedAt else { return false }
        return Date().timeIntervalSince(loadedAt) < maxCacheAgeSeconds
    }

    private func loadAd() {
        let unitID = remoteConfigManager.resolveAdUnitId(platform: .ios, format: .appOpen)
        isLoading = true
        print("[Ads][iOS][AppOpen] Loading ad unit=\(unitID)")
        AdDebugStatusStore.shared.updateLoad(format: "app_open", status: "loading")
        AppOpenAd.load(with: unitID, request: Request()) { [weak self] ad, error in
            Task { @MainActor [weak self] in
                guard let self else { return }
                self.isLoading = false
                if let error {
                    self.appOpenAd = nil
                    self.loadedAt = nil
                    print("[Ads][iOS][AppOpen] Load failed: \(error.localizedDescription)")
                    print("[Ads][iOS] Load fail for app_open: \(error.localizedDescription)")
                    AdDebugStatusStore.shared.updateLoad(format: "app_open", status: "failed:\(error.localizedDescription)")
                    return
                }
                self.appOpenAd = ad
                self.loadedAt = Date()
                self.appOpenAd?.fullScreenContentDelegate = self
                print("[Ads][iOS][AppOpen] Load succeeded.")
                print("[Ads][iOS] Load success for app_open")
                AdDebugStatusStore.shared.updateLoad(format: "app_open", status: "loaded")
                if self.wantsForegroundPresentation {
                    self.showIfAvailable()
                }
            }
        }
    }

    private func showIfAvailable() {
        guard remoteConfigManager.areAdsEnabled() else { return }
        guard remoteConfigManager.isAdFormatEnabled(format: .appOpen) else {
            AdDebugStatusStore.shared.updateShow(format: "app_open", status: "skipped:format_off")
            return
        }
        guard MobileAdsBootstrap.shared.isReady else {
            AdDebugStatusStore.shared.updateShow(format: "app_open", status: "waiting:sdk_init")
            preloadAdIfNeeded()
            return
        }
        guard !isShowing else { return }
        guard !presentationBlockedByAuthGate else {
            print("[Ads][iOS][AppOpen] Skipping show; auth gate is visible.")
            AdDebugStatusStore.shared.updateShow(format: "app_open", status: "skipped:auth_gate")
            return
        }
        guard !adManager.isShowingFullScreenAd else {
            print("[Ads][iOS][AppOpen] Skipping show; another fullscreen ad is already visible.")
            AdDebugStatusStore.shared.updateShow(format: "app_open", status: "skipped:fullscreen_visible")
            return
        }

        guard isAdFresh, let ad = appOpenAd else {
            print("[Ads][iOS][AppOpen] No fresh ad available; preloading.")
            AdDebugStatusStore.shared.updateShow(format: "app_open", status: "skipped:not_ready")
            preloadAdIfNeeded()
            return
        }

        guard let presenter = Self.topViewController() else {
            print("[Ads][iOS][AppOpen] No presenter view controller available.")
            scheduleRetryIfNeeded(reason: "presenter_missing")
            return
        }
        guard UIApplication.shared.applicationState == .active else {
            print("[Ads][iOS][AppOpen] Skipping show because application is not active.")
            return
        }
        guard !Self.hasPresentedModal() else {
            print("[Ads][iOS][AppOpen] Skipping show because another full-screen/modal UI is already presented.")
            scheduleRetryIfNeeded(reason: "modal_visible")
            return
        }

        isShowing = true
        wantsForegroundPresentation = false
        ad.fullScreenContentDelegate = self
        print("[Ads][iOS][AppOpen] Presenting app open ad.")
        ad.present(from: presenter)
    }

    private func scheduleRetryIfNeeded(reason: String) {
        guard wantsForegroundPresentation else { return }
        pendingShowRetryTask?.cancel()
        pendingShowRetryTask = Task { @MainActor [weak self] in
            guard let self else { return }
            try? await Task.sleep(for: .milliseconds(450))
            guard !Task.isCancelled else { return }
            print("[Ads][iOS][AppOpen] Retrying show after \(reason).")
            self.showIfAvailable()
        }
    }

    func adDidRecordImpression(_ ad: FullScreenPresentingAd) {
        print("[Ads][iOS][AppOpen] Impression recorded.")
        AdImpressionReporter.shared.report(
            adFormat: "app_open",
            adUnitId: remoteConfigManager.resolveAdUnitId(platform: .ios, format: .appOpen),
            placement: "app_open"
        )
        AdDebugStatusStore.shared.updateShow(format: "app_open", status: "impression")
    }

    func adWillPresentFullScreenContent(_ ad: FullScreenPresentingAd) {
        print("[Ads][iOS][AppOpen] Will present.")
        print("[Ads][iOS] Show success for app_open")
        AdDebugStatusStore.shared.updateShow(format: "app_open", status: "shown")
    }

    func adDidDismissFullScreenContent(_ ad: FullScreenPresentingAd) {
        print("[Ads][iOS][AppOpen] Dismissed.")
        pendingShowRetryTask?.cancel()
        pendingShowRetryTask = nil
        isShowing = false
        appOpenAd = nil
        loadedAt = nil
        AdDebugStatusStore.shared.updateShow(format: "app_open", status: "dismissed")
        preloadAdIfNeeded()
    }

    func ad(
        _ ad: FullScreenPresentingAd,
        didFailToPresentFullScreenContentWithError error: Error
    ) {
        print("[Ads][iOS][AppOpen] Failed to show: \(error.localizedDescription)")
        pendingShowRetryTask?.cancel()
        pendingShowRetryTask = nil
        isShowing = false
        appOpenAd = nil
        loadedAt = nil
        AdDebugStatusStore.shared.updateShow(format: "app_open", status: "failed:\(error.localizedDescription)")
        preloadAdIfNeeded()
    }

    @MainActor
    private static func topViewController(
        base: UIViewController? = nil
    ) -> UIViewController? {
        let resolvedBase = base ?? UIApplication.shared.connectedScenes
            .compactMap { $0 as? UIWindowScene }
            .flatMap { $0.windows }
            .first { $0.isKeyWindow }?
            .rootViewController
        if let nav = resolvedBase as? UINavigationController {
            return topViewController(base: nav.visibleViewController)
        }
        if let tab = resolvedBase as? UITabBarController, let selected = tab.selectedViewController {
            return topViewController(base: selected)
        }
        if let presented = resolvedBase?.presentedViewController {
            return topViewController(base: presented)
        }
        return resolvedBase
    }

    @MainActor
    private static func hasPresentedModal() -> Bool {
        let root = UIApplication.shared.connectedScenes
            .compactMap { $0 as? UIWindowScene }
            .flatMap { $0.windows }
            .first { $0.isKeyWindow }?
            .rootViewController
        return root?.presentedViewController != nil
    }
}
#elseif canImport(UIKit)
final class AppOpenAdManager {
    init(remoteConfigManager: RemoteConfigManager, adManager: AdManager) {
        _ = remoteConfigManager
        _ = adManager
    }

    func preloadAdIfNeeded() {}

    func sceneDidBecomeActive() {}

    func setPresentationBlockedByAuthGate(_ blocked: Bool) {
        _ = blocked
    }
}
#else
final class AppOpenAdManager {
    init(remoteConfigManager: RemoteConfigManager, adManager: AdManager) {
        _ = remoteConfigManager
        _ = adManager
    }

    func preloadAdIfNeeded() {}

    func sceneDidBecomeActive() {}

    func setPresentationBlockedByAuthGate(_ blocked: Bool) {
        _ = blocked
    }
}
#endif
